rfc/state-file.md

```diff
@@ -0,0 +1,94 @@
+like loadbalancer IPs or UID of the cluster.
+
+If this information is configurable by a user, and required to set up the cluster, the user saves it in Constellation's config file.
+Information that is not configurable, and that is generated by the cluster during the creation or init phase is either saved in Terraform output files,
+or in the cluster ID file (`constellation-id.json`).
+
+This creates problems since the relevant information may be split over multiple files, and it is not always clear at what point certain information is available in the cluster ID file.
+Additionally, commands like `upgrade` and `init` currently depend on Terraform files being present in the working directory,
+which may not be the case if the cloud resources were manually created.
+
+## Proposal
+
+Add a Constellation state file to keep track of this information.
+This file should replace the current `constellation-id.json` file and be the single source of truth for this kind of cluster relevant information.
+The file should keep track of all information currently kept in `constellation-id.json`, as well as information we currently have to read from Terraform files
+during `init` or `upgrade`.
+
+The file should be considered "read-only" by users, if Constellation's cloud resource were created using the Constellation CLI.
+If the cloud resources were created manually by a user, they will need to manually fill in some of the state file's fields.
+
+This will allow us to more easily decouple resource creation/management from Constellation.
+Meaning it should be much easier for Constellation users to have different parties create and update cloud resources and IAM using their own tooling (e.g. Terraform),
+and only use the Constellation CLI for Kubernetes cluster management (Kubernetes versions, service versions, image upgrades etc.).
+
+## Format
+
+The state file will be in either YAML or JSON format, and contain a version number to keep track of necessary migrations or breaking formats.
+The file should be split into multiple sections for each relevant resource.
+
+Information that is only used to upgrade Terraform resources (e.g. IAM) should not be part of this file.
+
+The following is a list of proposed sections:
+
+* Cloud resources
+
+    Information about resources we currently create using Terraform.
+    Essentially, we want to record the information we currently write to the cluster ID file during `create`,
+    as well as the Terraform output during `create`.
+
+    This field is expected to be manually set if cloud resources were manually deployed by a user.
+    If cloud resource were created with the Constellation CLI (using `constellation create`),
+    it will be populated automatically.
+
+* Runtime resources
+
+    cluster ID, measurement salt etc.
+    This field will be automatically populated during `constellation init`
+
+YAML file example:
+
+```yaml
+version: "v1"
+cloudResources:
```

Comment:
one thought: if only `cloudResources` are ever thought to be touched by the user shouldn't we let the user create a minimal file and have a separate file for `runtimeResources`?

Reply:
I would like to not have multiple files. My understanding is that part of the problem we want to address with the statefile is that we have multiple files.

I agree with the point on safety and feel like we should make it obvious to the user what can be changed and what not.